src/Common/LatencyBuckets.cpp

```diff
@@ -0,0 +1,107 @@
+    M(DiskS3FirstByteWriteAttempt2Microseconds, "Time of first byte write to DiskS3 storage (attempt 2).", 100, 1000, 10000, 100000, 300000, 500000, 1000000, 2000000, 5000000, 10000000, 15000000, 20000000, 25000000, 30000000, 35000000) \
+    M(DiskS3FirstByteReadAttemptNMicroseconds, "Time of first byte read from DiskS3 storage (attempt N).", 100, 1000, 10000, 100000, 300000, 500000, 1000000, 2000000, 5000000, 10000000, 15000000, 20000000, 25000000, 30000000, 35000000) \
+    M(DiskS3FirstByteWriteAttemptNMicroseconds, "Time of first byte write to DiskS3 storage (attempt N).", 100, 1000, 10000, 100000, 300000, 500000, 1000000, 2000000, 5000000, 10000000, 15000000, 20000000, 25000000, 30000000, 35000000) \
+    M(DiskS3ReadConnectMicroseconds, "Time to connect for GET and HEAD requests to DiskS3 storage.", 100, 1000, 10000, 100000, 200000, 300000, 500000, 1000000, 1500000) \
+    M(DiskS3WriteConnectMicroseconds, "Time to connect for POST, DELETE, PUT and PATCH requests to DiskS3 storage.", 100, 1000, 10000, 100000, 200000, 300000, 500000, 1000000, 1500000) \
+
+
+namespace LatencyBuckets
+{
+
+#define M(NAME, DOCUMENTATION, ...) extern const Event NAME = Event(__COUNTER__);
+    APPLY_FOR_EVENTS(M)
+#undef M
+
+constexpr Event END = Event(__COUNTER__);
+
+BucketList global_bucket_lists_array[END] =
+{
+#define M(NAME, DOCUMENTATION, ...) BucketList(CH_VA_ARGS_NARGS(__VA_ARGS__) + 1),
+    APPLY_FOR_EVENTS(M)
+#undef M
+};
+BucketLists global_bucket_lists(global_bucket_lists_array);
+
+const Event BucketLists::num_events = END;
+
+void BucketLists::reset()
+{
+    if (bucket_lists)
+    {
+        for (Event i = Event(0); i < num_events; ++i)
+            for (auto & bucket : bucket_lists[i])
+                bucket.store(0, std::memory_order_relaxed);
+    }
+}
+
+const char * getName(Event event)
+{
+    static const char * strings[] =
+    {
+    #define M(NAME, ...) #NAME,
+        APPLY_FOR_EVENTS(M)
+    #undef M
+    };
+
+    return strings[event];
+}
+
+const char * getDocumentation(Event event)
+{
+    static const char * strings[] =
+    {
+    #define M(NAME, DOCUMENTATION, ...) DOCUMENTATION "",
+        APPLY_FOR_EVENTS(M)
+    #undef M
+    };
+
+    return strings[event];
+}
+
+std::vector<Count> & getBucketBounds(Event event)
+{
+    static std::vector<Count> all_buckets_bounds[] =
+    {
+    #define M(NAME, DOCUMENTATION, ...) {__VA_ARGS__, UINT_MAX},
+        APPLY_FOR_EVENTS(M)
+    #undef M
+    };
+
+    return all_buckets_bounds[event];
+}
+
+Event end() { return END; }
+
+void increment(Event event, Count amount)
+{
+    global_bucket_lists.increment(event, amount);
+}
+
+void BucketLists::increment(Event event, Count amount)
+{
+    auto & bucket_bounds = getBucketBounds(event);
+    auto & bucket_list = this->bucket_lists[event];
+
+    for (ssize_t i = bucket_list.size() - 1; i > -1 && bucket_bounds[i] >= amount; --i)
```

Comment:
Lest make that call cheaper. It is better to do all the calculation about percentiles when we dump it to the table.